packages/core/strapi/ee/index.js

```diff
@@ -0,0 +1,218 @@
+  ee.licenseInfo = pick('licenseKey', ee.licenseInfo);
+  ee.enabled = false;
+};
+
+const readLicense = (directory) => {
+  try {
+    const path = join(directory, 'license.txt');
+    return fs.readFileSync(path).toString();
+  } catch (error) {
+    if (error.code !== 'ENOENT') {
+      // Permission denied, directory found instead of file, etc.
+    }
+  }
+};
+
+const fetchLicense = async (key, fallback) => {
+  const fallbackToStoredLicense = () => {
+    const error = 'Could not proceed to the online verification of your license.';
+
+    if (fallback) {
+      ee.logger(`${error} We will try to use the locally stored one as a potential fallback.`);
+      return fallback;
+    }
+
+    disable(`${error} Sorry for the inconvenience. Starting in CE.`);
+    return null;
+  };
+
+  try {
+    const response = await fetch(`https://license.strapi.io/api/licenses/${key}`);
+
+    if (response.status >= 500) {
+      return fallbackToStoredLicense();
+    }
+
+    if (response.status >= 400) {
+      disable();
+      return null;
+    }
+
+    return response.text();
+  } catch {
+    return fallbackToStoredLicense();
+  }
+};
+
+const verifyLicense = (license) => {
+  const [signature, base64Content] = Buffer.from(license, 'base64').toString().split('\n');
+  const stringifiedContent = Buffer.from(base64Content, 'base64').toString();
+
+  const verifier = crypto.createVerify('RSA-SHA256');
+  verifier.update(stringifiedContent);
+  verifier.end();
+
+  const verified = verifier.verify(publicKey, signature, 'base64');
+
+  return { verified, licenseInfo: verified ? JSON.parse(stringifiedContent) : null };
+};
+
+let initialized = false;
+
+const init = (licenseDir, logger) => {
+  // Can only be executed once, to prevent any abuse of the optimistic behavior
+  if (initialized) {
+    return;
+  }
+
+  initialized = true;
+  ee.logger = logger;
+
+  if (process.env.STRAPI_DISABLE_EE?.toLowerCase() === 'true') {
+    return;
+  }
+
+  const license = process.env.STRAPI_LICENSE || readLicense(licenseDir);
+
+  if (license) {
+    const { verified, licenseInfo } = verifyLicense(license);
+
+    if (verified) {
+      ee.enabled = true; // Optimistically enable EE during initialization
+      ee.licenseInfo = licenseInfo;
+    } else {
+      return disable();
+    }
+  }
+};
+
+const oneMinute = 1000 * 60;
+
+const onlineUpdate = async (db) => {
+  const transaction = await db.transaction();
+
+  try {
+    // TODO: Use the core store interface instead, it does not support transactions and "FOR UPDATE" at the moment
+    const eeInfo = await db
+      .queryBuilder(coreStoreModel.uid)
+      .where({ key: 'ee_information' })
+      .select('value')
+      .first()
+      .transacting(transaction)
+      .forUpdate()
+      .execute()
+      .then((result) => (result ? JSON.parse(result.value) : result));
+
+    const useStoredLicense = eeInfo?.lastOnlineCheck > Date.now() - oneMinute;
```

Comment:
It's long enough to only generate one outbound call at startup time while being short enough to allow the user to retry regularly in case of a license registry failure.

Though, we could make it longer though if you have a better duration in mind for sure.

Reply:
if the fetch failed you don't storage the date anyway so you will retry at every restart untill it succeeds right ? Not sure I see the usecase

Reply:
Data is stored in the database even in case of failure, so no we won't retry at every restart, I clarified that in the latest changes.